src/server/pfs/server/master.go

```diff
@@ -34,9 +42,122 @@ func (d *driver) master(ctx context.Context) {
 			gc := chunk.NewGC(d.storage.ChunkStorage())
 			return gc.RunForever(ctx)
 		})
+		eg.Go(func() error {
+			return d.finishCommits(ctx)
+		})
 		return eg.Wait()
 	}, backoff.NewInfiniteBackOff(), func(err error, _ time.Duration) error {
 		log.Errorf("error in pfs master: %v", err)
 		return nil
 	})
 }
+
+func (d *driver) finishCommits(ctx context.Context) error {
+	return d.commits.ReadOnly(ctx).WatchF(func(ev *watch.Event) error {
+		if ev.Type == watch.EventError {
+			return ev.Err
+		}
+		var key string
+		commitInfo := &pfs.CommitInfo{}
+		if err := ev.Unmarshal(&key, commitInfo); err != nil {
+			return err
+		}
+		if commitInfo.Finishing == nil || commitInfo.Finished != nil {
+			return nil
+		}
+		return backoff.RetryUntilCancel(ctx, func() error {
+			id, err := d.getFileSet(ctx, commitInfo.Commit)
+			if err != nil {
+				if pfsserver.IsCommitNotFoundErr(err) {
```

Comment:
I think we should check for non-transient errors in the other callback, below.  Everything in here returns errors, the second callback checks them for things we can't hope will disappear.

Reply:
Hmm, that seems a bit weird because we would have to bubble the error up and handle it outside of the function rather than just considering it a terminal condition local to the computation. Having the commit disappear is expected behavior for a squash commit.